serverkeyapi/internal/api.go

```diff
@@ -45,51 +79,153 @@ func (s *ServerKeyAPI) FetchKeys(
 	for _, fetcher := range s.OurKeyRing.KeyFetchers {
+		// If there are no more keys to look up then stop.
 		if len(requests) == 0 {
 			break
 		}
-		if fetcherResults, err := fetcher.FetchKeys(ctx, requests); err == nil {
-			// We successfully got some keys. Add them to the results and
-			// remove them from the request list.
+		fetcherCtx, fetcherCancel := context.WithTimeout(ctx, time.Second*30)
+		defer fetcherCancel()
+		logrus.WithFields(logrus.Fields{
+			"fetcher_name": fetcher.FetcherName(),
+		}).Infof("Fetching %d key(s)", len(requests))
+
+		if fetcherResults, err := fetcher.FetchKeys(fetcherCtx, requests); err == nil {
+			// Build a map of the results that we want to commit to the
+			// database. We do this in a separate map because otherwise we
+			// might end up trying to rewrite database entries.
+			storeResults := map[gomatrixserverlib.PublicKeyLookupRequest]gomatrixserverlib.PublicKeyLookupResult{}
+
+			// Now let's look at the results that we got from this fetcher.
 			for req, res := range fetcherResults {
-				if !res.WasValidAt(now, true) {
-					// We appear to be past the key validity. Don't return this
-					// key with the results.
-					continue
+				if prev, ok := results[req]; ok {
+					// We've already got a previous entry for this request
+					// so let's see if the newly retrieved one contains a more
+					// up-to-date validity period.
+					if res.ValidUntilTS > prev.ValidUntilTS {
+						// This key is newer than the one we had so let's store
+						// it in the database.
+						if req.ServerName != s.Cfg.Matrix.ServerName {
+							storeResults[req] = res
+						}
+					}
+				} else {
+					// We didn't already have a previous entry for this request
+					// so store it in the database anyway for now.
+					if req.ServerName != s.Cfg.Matrix.ServerName {
+						storeResults[req] = res
+					}
 				}
+
+				// Update the results map with this new result. If nothing
+				// else, we can try verifying against this key.
 				results[req] = res
-				delete(requests, req)
+
+				// If the key is valid right now then we can remove it from the
+				// request list as we won't need to re-fetch it.
+				if res.WasValidAt(now, true) {
+					delete(requests, req)
+				}
 			}
-			if err = s.OurKeyRing.KeyDatabase.StoreKeys(ctx, fetcherResults); err != nil {
+
+			// Store the keys from our store map.
+			if err = s.OurKeyRing.KeyDatabase.StoreKeys(ctx, storeResults); err != nil {
+				logrus.WithError(err).WithFields(logrus.Fields{
+					"fetcher_name":  fetcher.FetcherName(),
+					"database_name": s.OurKeyRing.KeyDatabase.FetcherName(),
+				}).Errorf("Failed to store keys in the database")
 				return nil, fmt.Errorf("server key API failed to store retrieved keys: %w", err)
 			}
+
+			if len(storeResults) > 0 {
+				logrus.WithFields(logrus.Fields{
+					"fetcher_name": fetcher.FetcherName(),
+				}).Infof("Updated %d of %d key(s) in database", len(storeResults), len(results))
+			}
+		} else {
```

Comment:
This is what I mean about the whole "10s of lines for if statements makes it hard to grok what the else part means". I have to literally carefully scan back up the page to try to find it, 52 lines above.

Reply:
I'd just invert then outdent, and then return (assuming you put the body of the for loop in a separate function) e.g

```go
if fetcherResults, err := fetcher.FetchKeys(fetcherCtx, requests); err != nil {
    logrus.WithError(err).WithFields(logrus.Fields{
        "fetcher_name": fetcher.FetcherName(),
    }).Errorf("Failed to retrieve %d key(s)", len(requests))
    return
}
// rest of body
```

Reply:
`FetchKeys` is broken up now.